docs/remote-attestation.md

```diff
@@ -0,0 +1,148 @@
+# Remote Attestation
+
+A crucial advantage of a Trusted Execution Environment (TEE) is the ability to
+perform **Remote Attestation**. It allows software to remotely check the TEE
+platform authenticity and also provides the remote software with the information
+about the code running inside the TEE. And if the code of the application is
+**reproducibly buildable**, the software can check that the TEE platform is
+running the application that is expected to run.
+
+One of the main components used in the remote attestation process is an
+**Attestation Report**, which is a data structure signed by the TEE platform and
+containing information identifying the code that is running inside the TEE. This
+report can be checked to confirm that it is signed by the **TEE Provider** (e.g.
+AMD or Intel), which results in evidence that the code is running on a genuine
+TEE platform.
+
+## Overview
+
+The Remote Attestation protocol is implemented as a simplification of the
+[Enclave Key Exchange Protocol (EKEP)](https://asylo.dev/docs/concepts/ekep.html)
+handshake protocol.
+
+The Remote Attestation protocol works on top of the [gRPC](https://grpc.io/)
+protocol and consists of 2 stages:
+
+1. Handshake
+   - Server proves its identity by sending a cryptographically signed
+     Attestation Report
+   - Both Client and Server establish an encrypted connection with keys that are
+     bound to the TEE platform's firmware
+1. Data Exchange
+   - The keys generated during the Handshake are used to encrypt data sent
+     between Client and Server
+
+Prior to the _Handshake_ both Client and Server each generate an individual
+[ECDSA-P256](https://datatracker.ietf.org/doc/html/rfc6979) key pair called
+**Signing Key Pair**. These keys are persistent between connections (but
+regenerated at startup) and are used to sign **Transcripts**: hashes of previous
+handshake messages that are sent in each new handshake message to prevent
+[Replay Attacks](https://en.wikipedia.org/wiki/Replay_attack). Each handshake
+message also includes a random string so that transcripts always contain new
+random values.
+
+Also for each individual connection both Client and Server each generate an
+[X25519](https://datatracker.ietf.org/doc/html/rfc7748) Diffie-Hellman key pair
+called **Ephemeral Key Pair**. These keys are used to establish a shared secret
+between Client and Server.
+
+The shared secret is used to generate 2 shared
+[AES-256-GCM](https://datatracker.ietf.org/doc/html/rfc5288) **Session Keys**
+that are used during the _Data Exchange_ stage:
+
+- Server Session Key
+  - Which is used to encrypt messages sent by the Server
+- Client Session Key
+  - Which is used to encrypt messages sent by the Client
+
+It's important to node that for each new request **Client** performs the Remote
+Attestation from the start in order to create a new pair of Session keys.
```

Comment:
Aren't these just like the same key? The client and server both have a copy, but of the same key.

Reply:
These are different keys `K_s` and `K_c`.
But Client and server have both of them.

`K_s` is used by the Server to encrypt outgoing messages, and by the Client to decrypt incoming messages.
`K_c` is used by the Client to encrypt outgoing messages, and by the Server to decrypt incoming messages.

This is necessary to prevent the attacker from sending the message back to the sender.